### PR TITLE
#223: withAuthUserTokenAPI

### PR DESCRIPTION
**Pull request by gavinharris-dev**
Introduced an API Middleware function which can be used to wrap Next JS API route handlers.

I'm sorry, I am not impressed with my test file. If I get time today I will review and rework to try and DRY up a little.

### CODE REVIEW

File: src/withAuthUserTokenAPI.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    res.status(401).send({ message: 'Unauthorized', code: 1 })
+    return
+  }
+
+  const token = req.headers.authorization.split(' ')[1]
+
+  try {
+    const authUser = await verifyIdToken(token)
+
+    if (!authUser.id) {
```
Comment (gavinharris-dev) [inline src/withAuthUserTokenAPI.js:29]: I think that I am not getting 100% how `verifyIdToken` works here. I assumed (when creating) that this method could possibly return a new user (with a null `id`); however I seem to not be able to get this to happen; @kmjennison are you able to cast your eye on these lines and provide some insight?

Comment (alextaymx) [reply src/withAuthUserTokenAPI.js:29]: Hi , I'm a newbie for this library , is this related to getting AuthUser in next js api route handler?

Comment (gavinharris-dev) [reply src/withAuthUserTokenAPI.js:29]: So I've 'fixed' the test to hit the 2 lines that I was missing. 

However I am still unconvinced that the scenario that I am testing for is possible (with this usage of the `verifyToken` method). From what I can understand (which is limited) the `verifyIdToken` will throw and error if the ID Token is not valid and so there _should_ be no way that I get an `AuthUser` that does not have an `id` property set. 

Am I missing something?

**Comment by gavinharris-dev** [comment]
Hi Alex,

Yes the method withAuthUserTokenAPIis a NextJS API Middleware which would
be used to wrap up API route handlers. The route handler's request object
will then (if Auth'n is successful) be populated with an AuthUser property.

Does that make sense?
Gavin

**Comment by kmjennison** [comment]
Thanks for putting work into this PR.

In my view, we should leave this functionality up to the app developer. It feels out of scope for this package, which is aimed at easily using a standardized AuthUser, not necessarily prescribing _how_ to use the AuthUser. I think `verifyIdToken` and the proposal in #223 cover the functionality needed for accessing the AuthUser in API routes.

One of my concerns is that this middleware is very opinionated: it expects the token in a specific header, it only checks for the `id` property to be set (not, for example, email address), and it returns a specific response. Those are all things that different apps might want to customize. Where the current API is too opinionated, I'm hoping to generalize it (for example, #233).

I'm open to your thoughts if you feel I'm missing something, but I'm inclined to close this PR. Sorry, I know it's frustrating—I'll update our contributing guidelines to ask contributors to open issues for discussion before doing the work.

**Comment by gavinharris-dev** [comment]
All good @kmjennison; I will need this API Middleware for my existing project and so there is no time lost (also good learning how Middleware is to work on NextJS API routes).

With respect to the opinionated nature of the API; you are correct, but the opinions are not 100% of the Middleware and more how Authorization is handled in REST API's:

* The HTTP Header utilized is a HTTP standard (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization).
* The HTTP Response (HTTP Status 401) is also standard (https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401)
* The validation on the ID Token confuses me a little (trying to get my head around how `verifyIdToken` would respond in this scenario). I used the [withAuthUserTokenSSR](https://github.com/gladly-team/next-firebase-auth/blob/main/src/withAuthUserTokenSSR.js#L95) check but not really sure if this is valid.

Anyway, happy to put together an RFC and have some discussion (should have done this in the first place!) on the nature of this API Middleware and how we could go about allowing configuration hooks into the process.

**Comment by kmjennison** [comment]
@gavinharris-dev Sounds good, thanks! At the very least, an example of middleware in the docs here would be excellent.

**Comment by gavinharris-dev** [comment]
I will create an example project showing this use case for this Middleware.

**Comment by alextaymx** [comment]
Hi @gavinharris-dev , is this available in the latest npm package?

**Comment by gavinharris-dev** [comment]
Hi @alextaymx; this will not be available as yet. I need to put together an RFC and think about all the use cases that this will resolve (and if said use cases are actually something that this project should resolve!).

You are more then welcome to jump in and grab the code and hold it locally (look at https://github.com/gavinharris-dev/next-firebase-auth/blob/fe7dca8f39d4b7b4e2369c15150c5d8ebcc87b6f/src/withAuthUserTokenAPI.js). If it is useful then please comment on the RFC (when I get time to write it up!)

**Comment by gavinharris-dev** [comment]
> Hi @gavinharris-dev , is this available in the latest npm package?
> 
> Thanks for developing this feature, I was wondering how to get AuthUser in api routes without doing the hard job of putting authorization header and checking it every time.

FYI: https://gist.github.com/gavinharris-dev/0d936daa3d2def14134c80767dc998a2

**Pull request closed without merging**